## GPS Accuracy Using Root Mean Squared Error (RMSE)

In [245]:
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd
import math

data = pd.read_excel('gps_data.xlsx')
lat = data['Latitude']
lon = data['Longitude']
true_lat = data['True Latitude']
true_lon = data['True Longitude']
true_dist1, true_dist2 = [], []
dist1, dist2 = [], []
mse1, mse2 = [], []
loc_rmse = []

dlat = np.deg2rad(lat - true_lat)
dlon = np.deg2rad(lon - true_lon)
lat = np.deg2rad(lat)
lon = np.deg2rad(lon)
true_lat = np.deg2rad(true_lat)
true_lon = np.deg2rad(true_lon)

for i in lat:
    true_dist_value = 0
    true_dist.append(true_dist_value)

# Haversine Formula
rEarth = 6371000
a = (np.sin(dlat / 2) ** 2) + np.cos(true_lat) * np.cos(lat) * ((np.sin(dlon / 2) / 2) ** 2)
c = 2 * np.arcsin(np.sqrt(a))
dist = rEarth * c

### First Location (JP Rizal Street, Brgy, Pililla - Jala-Jala - Pakil Rd, Pililla, 1910 Rizal)

In [228]:
data[0:30]

Latitude   Longitude  True Latitude  True Longitude  \
0   14.430120  121.337312      14.430289      121.337339   
1   14.430121  121.337312      14.430289      121.337339   
2   14.430121  121.337309      14.430289      121.337339   
3   14.430124  121.337305      14.430289      121.337339   
4   14.430125  121.337304      14.430289      121.337339   
5   14.430121  121.337299      14.430289      121.337339   
6   14.430120  121.337293      14.430289      121.337339   
7   14.430117  121.337290      14.430289      121.337339   
8   14.430118  121.337291      14.430289      121.337339   
9   14.430119  121.337291      14.430289      121.337339   
10  14.430119  121.337289      14.430289      121.337339   
11  14.430120  121.337289      14.430289      121.337339   
12  14.430120  121.337291      14.430289      121.337339   
13  14.430120  121.337292      14.430289      121.337339   
14  14.430125  121.337295      14.430289      121.337339   
15  14.430130  121.337297      14.430289      121.337339   
16  14.430131  121.337296      14.430289      121.337339   
17  14.430133  121.337298      14.430289      121.337339   
18  14.430134  121.337301      14.430289      121.337339   
19  14.430134  121.337304      14.430289      121.337339   
20  14.430132  121.337302      14.430289      121.337339   
21  14.430130  121.337301      14.430289      121.337339   
22  14.430127  121.337300      14.430289      121.337339   
23  14.430126  121.337299      14.430289      121.337339   
24  14.430120  121.337299      14.430289      121.337339   
25  14.430116  121.337299      14.430289      121.337339   
26  14.430115  121.337304      14.430289      121.337339   
27  14.430120  121.337310      14.430289      121.337339   
28  14.430120  121.337313      14.430289      121.337339   
29  14.430122  121.337316      14.430289      121.337339   

                                             Location  
0   JP Rizal Street, Brgy, Pililla - Jala-Jala - P...  
1                                                 NaN  
2                                                 NaN  
3                                                 NaN  
4                                                 NaN  
5                                                 NaN  
6                                                 NaN  
7                                                 NaN  
8                                                 NaN  
9                                                 NaN  
10                                                NaN  
11                                                NaN  
12                                                NaN  
13                                                NaN  
14                                                NaN  
15                                                NaN  
16                                                NaN  
17                                                NaN  
18                                                NaN  
19                                                NaN  
20                                                NaN  
21                                                NaN  
22                                                NaN  
23                                                NaN  
24                                                NaN  
25                                                NaN  
26                                                NaN  
27                                                NaN  
28                                                NaN  
29                                                NaN

### First Location RMSE

In [246]:
dist1 = dist[0:30]
true_dist1 = true_dist[0:30]
mse1 = mean_squared_error(dist1, true_dist1)
np.sqrt(mse1)
print(np.sqrt(mse1))
loc_rmse.append(np.sqrt(mse1))

18.521240540130293


### Second Location (147 C. Tolentino Compound Banaag Street Pineda Pasig City)

In [230]:
data[30:60]

Latitude   Longitude  True Latitude  True Longitude  \
30  14.566052  121.058144       14.56603       121.05823   
31  14.566055  121.058136       14.56603       121.05823   
32  14.566056  121.058136       14.56603       121.05823   
33  14.566058  121.058128       14.56603       121.05823   
34  14.566060  121.058128       14.56603       121.05823   
35  14.566063  121.058121       14.56603       121.05823   
36  14.566065  121.058121       14.56603       121.05823   
37  14.566067  121.058121       14.56603       121.05823   
38  14.566068  121.058121       14.56603       121.05823   
39  14.566068  121.058121       14.56603       121.05823   
40  14.566065  121.058121       14.56603       121.05823   
41  14.566061  121.058128       14.56603       121.05823   
42  14.566059  121.058128       14.56603       121.05823   
43  14.566057  121.058128       14.56603       121.05823   
44  14.566057  121.058128       14.56603       121.05823   
45  14.566057  121.058128       14.56603       121.05823   
46  14.566055  121.058128       14.56603       121.05823   
47  14.566055  121.058121       14.56603       121.05823   
48  14.566059  121.058121       14.56603       121.05823   
49  14.566061  121.058121       14.56603       121.05823   
50  14.566063  121.058128       14.56603       121.05823   
51  14.566064  121.058128       14.56603       121.05823   
52  14.566068  121.058128       14.56603       121.05823   
53  14.566067  121.058128       14.56603       121.05823   
54  14.566067  121.058128       14.56603       121.05823   
55  14.566067  121.058128       14.56603       121.05823   
56  14.566066  121.058121       14.56603       121.05823   
57  14.566067  121.058113       14.56603       121.05823   
58  14.566069  121.058105       14.56603       121.05823   
59  14.566073  121.058098       14.56603       121.05823   

                                             Location  
30  147 C. Tolentino Compound Banaag Street Pineda...  
31                                                NaN  
32                                                NaN  
33                                                NaN  
34                                                NaN  
35                                                NaN  
36                                                NaN  
37                                                NaN  
38                                                NaN  
39                                                NaN  
40                                                NaN  
41                                                NaN  
42                                                NaN  
43                                                NaN  
44                                                NaN  
45                                                NaN  
46                                                NaN  
47                                                NaN  
48                                                NaN  
49                                                NaN  
50                                                NaN  
51                                                NaN  
52                                                NaN  
53                                                NaN  
54                                                NaN  
55                                                NaN  
56                                                NaN  
57                                                NaN  
58                                                NaN  
59                                                NaN

In [247]:
dist2 = dist[30:60]
true_dist2 = true_dist[30:60]
mse2 = mean_squared_error(dist2, true_dist2)
np.sqrt(mse2)
print(np.sqrt(mse2))
loc_rmse.append(np.sqrt(mse2))

6.791612817518172


## Remarks

The RMSE serves as a guide to determine the accuracy of a GPS within a specific range. It tells the observer that their true location might be within *n* meters of the true location (*n* is the RMSE).

Based on the data gathered by the researchers, they have noticed that the RMSE calculated on the first location is **18.52 meters**, this suggests that the true value of the coordinate lies within that range. A factor that could have caused the low accuracy are the presence of trees in the area. On the other hand, the accuracy on the second location is within **6.79 meters**. This is a normal range for a low-cost commercial GPS, however, according to the datasheet the NEO-M8N has an accuracy within 2.5 meters (under optimal conditions). It might be good to note that this was tested in a residential area (more reading should be done to understand the reasons why GPS signals are weakened).